xnmt/decoder.py

```diff
@@ -47,49 +50,52 @@ class MlpSoftmaxDecoder(RnnDecoder, Serializable):
-               mlp_hidden_dim=None, trg_embed_dim=None, dropout=None,
-               rnn_spec="lstm", residual_to_output=False, input_feeding=True,
-               param_init_lstm=None, param_init_context=None, bias_init_context=None,
+  def __init__(self, exp_global=Ref(Path("exp_global")),
+               rnn_layer=bare(UniLSTMSeqTransducer), input_dim=None,
+               mlp_hidden_dim=None, trg_embed_dim=None, input_feeding=True,
+               param_init_context=None, bias_init_context=None,
                param_init_output=None, bias_init_output=None,
                bridge=bare(CopyBridge), label_smoothing=0.0,
                vocab_projector=None, vocab_size = None, vocab = None,
                trg_reader = Ref(path=Path("model.trg_reader"), required=False)):
+               # obsolete or moved into rnn_layer:
+               #layers=1, lstm_dim=None,
+               #dropout=None,
+               #rnn_spec="lstm", residual_to_output=False,
+               #param_init_lstm=None,
     register_handler(self)
     self.param_col = exp_global.dynet_param_collection.param_col
     # Define dim
-    lstm_dim       = lstm_dim or exp_global.default_layer_dim
     self.mlp_hidden_dim = mlp_hidden_dim = mlp_hidden_dim or exp_global.default_layer_dim
     trg_embed_dim  = trg_embed_dim or exp_global.default_layer_dim
     input_dim      = input_dim or exp_global.default_layer_dim
     self.input_dim = input_dim
     self.label_smoothing = label_smoothing
     # Input feeding
     self.input_feeding = input_feeding
-    self.lstm_dim = lstm_dim
-    lstm_input = trg_embed_dim
+    rnn_input_dim = trg_embed_dim
     if input_feeding:
-      lstm_input += input_dim
+      rnn_input_dim += input_dim
+    assert rnn_input_dim == rnn_layer.input_dim, "Wrong input dimension in RNN layer"
     # Bridge
-    self.lstm_layers = layers
     self.bridge = bridge
 
     # LSTM
-    self.fwd_lstm  = RnnDecoder.rnn_from_spec(spec       = rnn_spec,
-                                              num_layers = layers,
-                                              input_dim  = lstm_input,
-                                              hidden_dim = lstm_dim,
-                                              model = self.param_col,
-                                              residual_to_output = residual_to_output)
-    param_init_lstm = param_init_lstm or exp_global.param_init
-    if not isinstance(param_init_lstm, GlorotInitializer): raise NotImplementedError("For the decoder LSTM, only Glorot initialization is currently supported")
-    if getattr(param_init_lstm,"gain",1.0) != 1.0:
-      for l in range(layers):
-        for i in [0,1]:
-          self.fwd_lstm.param_collection().parameters_list()[3*l+i].scale(param_init_lstm.gain)
-      
+    self.rnn_layer = rnn_layer
+
+    #=== TODO: what does this do and why is it needed? how to preserve it with
+    #===       the refactoring?
```

Comment:
You can safely delete this. The DyNet RNN builders are hard to configure regarding parameter initialization, so this was a kind of workaround, but with the XNMT builders parameter initialization is fully supported and already implemented.